pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentStickyKeyDispatcherMultipleConsumersClassic.java

```diff
@@ -560,8 +593,11 @@ public boolean hasSameKeySharedPolicy(KeySharedMeta ksm) {
                 && ksm.isAllowOutOfOrderDelivery() == this.allowOutOfOrderDelivery);
     }
 
-    public LinkedHashMap<Consumer, Position> getRecentlyJoinedConsumers() {
-        return recentlyJoinedConsumers;
+    public synchronized LinkedHashMap<Consumer, Position> getRecentlyJoinedConsumers() {
+        if (recentlyJoinedConsumers == null) {
+            return null;
+        }
+        return new LinkedHashMap<>(recentlyJoinedConsumers);
```

Comment:
why need to recreate a new `LinkedHashMap`

Reply:
Because it will return the original collection, which may be modified by other threads

Reply:
We should return a unmodified map instead of a new copy of the map?

Reply:
> We should return a unmodified map instead of a new copy of the map?

No, the thread that gets the unmodified map may encounter a `ConcurrentModificationException`, see following case:

- `thread-1`: get an unmodified map
  - call `map.foreach` 
- `thread-2`: get the lock and remove an item from the map.
- `thread-1` will get a `ConcurrentModificationException` because its `foreach` is in progress